WordPress/src/main/java/org/wordpress/android/ui/reader/usecases/ReaderSiteNotificationsUseCase.kt

```diff
@@ -0,0 +1,113 @@
+package org.wordpress.android.ui.reader.usecases
+
+import org.greenrobot.eventbus.Subscribe
+import org.greenrobot.eventbus.ThreadMode
+import org.wordpress.android.analytics.AnalyticsTracker.Stat.FOLLOWED_BLOG_NOTIFICATIONS_READER_MENU_OFF
+import org.wordpress.android.analytics.AnalyticsTracker.Stat.FOLLOWED_BLOG_NOTIFICATIONS_READER_MENU_ON
+import org.wordpress.android.datasets.ReaderBlogTableWrapper
+import org.wordpress.android.fluxc.Dispatcher
+import org.wordpress.android.fluxc.generated.AccountActionBuilder
+import org.wordpress.android.fluxc.store.AccountStore.AddOrDeleteSubscriptionPayload
+import org.wordpress.android.fluxc.store.AccountStore.AddOrDeleteSubscriptionPayload.SubscriptionAction.DELETE
+import org.wordpress.android.fluxc.store.AccountStore.AddOrDeleteSubscriptionPayload.SubscriptionAction.NEW
+import org.wordpress.android.fluxc.store.AccountStore.OnSubscriptionUpdated
+import org.wordpress.android.ui.reader.usecases.ReaderSiteNotificationsUseCase.SiteNotificationState.Failed.NoNetwork
+import org.wordpress.android.ui.reader.usecases.ReaderSiteNotificationsUseCase.SiteNotificationState.Failed.RequestFailed
+import org.wordpress.android.util.AppLog
+import org.wordpress.android.util.AppLog.T.API
+import org.wordpress.android.util.NetworkUtilsWrapper
+import org.wordpress.android.util.analytics.AnalyticsUtilsWrapper
+import javax.inject.Inject
+import kotlin.coroutines.Continuation
+import kotlin.coroutines.resume
+import kotlin.coroutines.suspendCoroutine
+
+/**
+ * This class handles reader notification events.
+ */
+class ReaderSiteNotificationsUseCase @Inject constructor(
+    private val dispatcher: Dispatcher,
+    private val analyticsUtilsWrapper: AnalyticsUtilsWrapper,
+    private val readerBlogTableWrapper: ReaderBlogTableWrapper,
+    private val networkUtilsWrapper: NetworkUtilsWrapper
+) {
+    private var continuation: Continuation<Boolean>? = null
+
+    suspend fun toggleNotification(blogId: Long): SiteNotificationState? {
+        if (continuation != null) {
+            // Toggling notification for multiple sites in parallel isn't supported
+            // as the user would lose the ability to undo the action
+            return null
+        }
+        if (!networkUtilsWrapper.isNetworkAvailable()) {
+            return NoNetwork
+        }
+
+        // We want to track the action no matter the result
+        trackEvent(blogId)
+
+        val succeeded = suspendCoroutine<Boolean> { cont ->
+            continuation = cont
+            updateSubscription(blogId)
+        }
+
+        if (succeeded) {
+            updateBlogInDb(blogId)
+            dispatcher.dispatch(AccountActionBuilder.newFetchSubscriptionsAction())
+        } else {
+            return RequestFailed
+        }
+        return null
```

Comment:
We might want to consider creating states we don't need to re-act to instead of returning null. When we return null for these states, the UseCase technically plays part in how and to what we'll react and we are trying to remove this "responsibility" from the UseCase. Moreover, I believe avoiding nulls wherever possible is always a win. Wdyt?

Reply:
Removed null -> 4a6aee5, added test case -> 18ec384